vertx-mongo-client/src/main/java/io/vertx/ext/mongo/impl/MongoClientImpl.java

```diff
@@ -651,44 +635,31 @@ public MongoClient dropIndex(String collection, String indexName, Handler<AsyncR
 
   @Override
   public io.vertx.ext.mongo.MongoClient distinctWithQuery(String collection, String fieldName, String resultClassname, JsonObject query, Handler<AsyncResult<JsonArray>> resultHandler) {
-    DistinctIterable distinctValues = findDistinctValuesWithQuery(collection, fieldName, resultClassname, query, resultHandler);
-
-    if (distinctValues != null) {
-      convertMongoIterable(distinctValues, resultHandler);
+    try {
+      DistinctIterable distinctValues = findDistinctValuesWithQuery(collection, fieldName, resultClassname, query);
+      if (distinctValues != null) {
+        convertMongoIterable(distinctValues, resultHandler);
+      }
+    } catch (ClassNotFoundException e) {
+      resultHandler.handle(Future.failedFuture(e));
     }
     return this;
   }
 
   @Override
-  public io.vertx.ext.mongo.MongoClient distinctBatch(String collection, String fieldName, String resultClassname, Handler<AsyncResult<JsonObject>> resultHandler) {
-    return distinctBatchWithQuery(collection, fieldName, resultClassname, new JsonObject(), resultHandler);
+  public ReadStream<JsonObject> distinctBatch(String collection, String fieldName, String resultClassname) {
+    return distinctBatchWithQuery(collection, fieldName, resultClassname, new JsonObject());
   }
 
   @Override
-  public io.vertx.ext.mongo.MongoClient distinctBatchWithQuery(String collection, String fieldName, String resultClassname, JsonObject query, Handler<AsyncResult<JsonObject>> resultHandler) {
-    DistinctIterable distinctValues = findDistinctValuesWithQuery(collection, fieldName, resultClassname, query, resultHandler);
-
-    if (distinctValues != null) {
-      Context context = vertx.getOrCreateContext();
-      Block valueBlock = value -> {
-        context.runOnContext(v -> {
-          Map mapValue = new HashMap();
-          mapValue.put(fieldName, value);
-          resultHandler.handle(Future.succeededFuture(new JsonObject(mapValue)));
-        });
-      };
-      SingleResultCallback<Void> callbackWhenFinished = (result, throwable) -> {
-        if (throwable != null) {
-          resultHandler.handle(Future.failedFuture(throwable));
-        }
-      };
-      try {
-        distinctValues.forEach(valueBlock, callbackWhenFinished);
-      } catch (Exception unhandledEx) {
-        resultHandler.handle(Future.failedFuture(unhandledEx));
-      }
+  public ReadStream<JsonObject> distinctBatchWithQuery(String collection, String fieldName, String resultClassname, JsonObject query) {
+    try {
+      MongoIterable<JsonObject> distinctValues = findDistinctValuesWithQuery(collection, fieldName, resultClassname, query)
+        .map(value -> new JsonObject().put(fieldName, value));
+      return new MongoIterableStream(vertx.getOrCreateContext(), distinctValues);
+    } catch (ClassNotFoundException e) {
```

Comment:
Wondering about Java 9

Reply:
What do you mean?

Reply:
We are loading a class using the current classloader. So it should work with Java 9 (module-less). However, it won't work if you attempt to load a class from another module. I guess it's fine (it's the module semantic), it just may be confusing.